File: ThreadSafeDelayedEvalution/ThreadSafeDelayedEvalution/Node.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ThreadSafeDelayedEvalution
{
    public class Node
    {

        private bool _dirty = false;
        private int _param = 0;
        private Node _parent = null;
        private bool _enabled = true;
        private int _result = 0;
        private List<Node> _observerList = new();

        private uint _refCounter = 0;


        private void CustomDeStructor()
        {

            //delete my self from observer list of parent if i was destroyed
            if (_parent._observerList.Contains(this))
            {
                _observerList.Remove(this);
            }
        }


        private void checkIfValidNode()
        {
            if (_observerList.Count == 0 && _refCounter == 0)
            {
                throw new AccessViolationException("This node is not valid and should never be called :(");
            }
        }



        public void  addReference()
        {

            //node is always invalid with this example code=?
            //checkIfValidNode();
            
            _refCounter++;
        }

        public void releaseReference()
        {
            checkIfValidNode();

            _refCounter--;
            if (_refCounter == 0 && _observerList.Count == 0)
            {

                CustomDeStructor();
            }

        }

        public void setParent(Node parent)
        {
            checkIfValidNode();

            _parent = parent;

            _parent.registerNotification(this);
            setDirty();
        }



/*
        public void setEnabled(bool enabled)
        {
            _enabled = enabled;
            setDirty();
        }*/

        public void setParam(int param)
        {
            checkIfValidNode();
            _param = param;
            setDirty();
        }

        // Return the sum of the parameter and the result of the parent:
        public int getResult()
        {
            checkIfValidNode();

            if (_dirty)
            {
                _dirty = false;
                _result = 0;
                if (_enabled)
                {
                    _result = _param;
                    if (_parent != null)
                    {
                        _result += _parent.getResult();
                        _parent.registerNotification(this);
                    }
                }
            }
            return _result;
        }

        public void setDirty()
        {
            checkIfValidNode();

            _dirty = true;

            foreach (var observer in _observerList)
            {
                observer.setDirty();
            }

            //TODO:: Clear i echt die Observer wenn is dirty gsetzt hab=?

            _observerList.Clear();
        }


        void registerNotification(Node depNode)
        {
            checkIfValidNode();

            if (!_observerList.Contains(depNode))
            {
                _observerList.Add(depNode);
            }
        }
    }
}

File: ThreadSafeDelayedEvaluationCPP/ThreadSafeDelayedEvaluationCPP/Node.cpp
#include "Node.h"

Node::~Node()
{

	const std::lock_guard<std::mutex>locked(_mutex);
	if(_parent != nullptr)
	{
		_parent->unsubscribeNotification(this);
		_parent->releaseReference();
		_parent = nullptr;
	}


}

void Node::addReference()
{
	const std::lock_guard<std::mutex>locked(_mutexRef);
	_refCounter++;
}

void Node::releaseReference()
{

	//set scope
	{
		const std::lock_guard<std::mutex>locked(_mutexRef);
		_refCounter--;
	}

	if(_refCounter == 0 && _observerList.empty())
	{
		delete this;
	}
}

void Node::setParent(Node* parent)
{
	const std::lock_guard<std::mutex>locked(_mutex);

	if(_parent != nullptr)
	{
		parent->unsubscribeNotification(this);
	}
	_parent = parent;

	_parent->registerNotification(this);
	setDirty();
}

void Node::setParam(int param)
{

	{
		_param = param;

	}
	setDirty();
}

int Node::getResult()
{
	if(_dirty)
	{
		_dirty = false;
		_result = 0;
		if(_enabled)
		{
			_result = _param;
			if(_parent != nullptr)
			{
				_result += _parent->getResult();
				_parent->registerNotification(this);
			}
		}
	}

	return _result;
}

void Node::setDirty()
{
	_dirty = true;
	const auto observerCopy = _observerList;
	_observerList.clear();
	for (const auto observer : observerCopy)
	{
		observer->setDirty();
	}

	_parent->unsubscribeNotification(this);
}

void Node::registerNotification(Node* depNode)
{

	const std::lock_guard<std::mutex>locked(_mutex);

	if(std::find(_observerList.begin(), _observerList.end(), depNode) != _observerList.end())
	{
		_observerList.push_back(depNode);
	}
}

void Node::unsubscribeNotification(const Node* depNode)
{
	const std::lock_guard<std::mutex>locked(_mutex);

	const auto childNode = std::find(_observerList.begin(), _observerList.end(), depNode);
	if(childNode!= _observerList.end())
	{
		_observerList.erase(childNode);
	}
}
